hack/generated/controllers/generic_controller.go

```diff
@@ -0,0 +1,749 @@
+			}
+
+			annotations := make(map[string]string)
+			annotations[ResourceStateAnnotation] = string(armclient.DeletingProvisioningState)
+			addAnnotations(metaObj, annotations)
+		} else {
+			controllerutil.RemoveFinalizer(mutMetaObject, GenericControllerFinalizer)
+		}
+
+		// TODO: Not exactly sure what this was doing
+		//if err := gr.Converter.FromResource(resource, mutMetaObject); err != nil {
+		//	return errors.Wrapf(err, "error gr.Converter.FromResource")
+		//}
+
+		return nil
+	}); err != nil && !apierrors.IsNotFound(err) {
+		return ctrl.Result{}, errors.Wrap(err, "failed to patch after starting delete")
+	}
+
+	// delete has started, check back to seen when the finalizer can be removed
+	return ctrl.Result{
+		RequeueAfter: 5 * time.Second,
+	}, nil
+}
+
+// updateFromNonTerminalDeleteState will call Azure to check if the resource still exists. If so, it will requeue, else,
+// the finalizer will be removed.
+func (gr *GenericReconciler) updateFromNonTerminalDeleteState(
+	ctx context.Context,
+	log logr.Logger,
+	resource genruntime.ArmResource,
+	metaObj genruntime.MetaObject) (ctrl.Result, error) {
+
+	// already deleting, just check to see if it still exists and if it's gone, remove finalizer
+	found, err := gr.ARMClient.HeadResource(ctx, resource.GetId(), resource.Spec().GetApiVersion())
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to head resource")
+	}
+
+	if found {
+		log.V(0).Info("Found resource: continuing to wait for deletion...")
+		return ctrl.Result{RequeueAfter: 10 * time.Second}, nil
+	}
+
+	err = patcher(ctx, gr.Client, metaObj, func(ctx context.Context, mutMetaObject genruntime.MetaObject) error {
+		controllerutil.RemoveFinalizer(mutMetaObject, GenericControllerFinalizer)
+		return nil
+	})
+
+	// patcher will try to fetch the object after patching, so ignore not found errors
+	if apierrors.IsNotFound(err) {
+		return ctrl.Result{}, nil
+	}
+	return ctrl.Result{}, err
+}
+
+func (gr *GenericReconciler) getStatus(ctx context.Context, id string, data *ReconcileMetadata) (genruntime.ArmTransformer, error) {
+	_, typedArmSpec, err := ResourceSpecToArmResourceSpec(gr, data.metaObj)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: do we tolerate not exists here?
+	armStatus, err := NewEmptyArmResourceStatus(data.metaObj)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to construct empty ARM status object for resource with ID: %q", id)
+	}
+
+	// Get the resource
+	err = gr.ARMClient.GetResource(ctx, id, typedArmSpec.GetApiVersion(), armStatus)
+	if data.log.V(0).Enabled() { // TODO: CHange this level
+		statusBytes, err := json.Marshal(armStatus)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to serialize ARM status to JSON for debugging")
+		}
+		data.log.V(0).Info("Got ARM status", "status", string(statusBytes))
+	}
+
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get resource with ID: %q", id)
+	}
```

Comment:
Is it valid to log if an error occurred?

Reply:
All of the errors returned by this function end up bubbling up to the top of the loop and producing an error log - so we're sorta already logging an error.

Not totally clear to me though if you are advocating that we log an error here or that we stop logging errors in the caller?

FWIW in these "fatal" cases (where we're aborting execution and returning an error), I don't think we need to put a log here since we already have the error being returned and the caller can log that if it decides it wants to (which right now we are always doing)

Reply:
I was thinking that it would make sense to log here as well as returning the error, but it appears the logging is already present elsewhere, so no change needed here.